src/api/api.ts

```diff
@@ -1,69 +1,175 @@
-    }
-    return response.json() as ResponsePayload;
+    return makeWebRequest<ResponsePayload, RequestPayload>(method, path, payload, responseParser);
   } else {
-    const response = await tauriFetch(REFSTUDIO_HOST + path, {
-      method,
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-explicit-any
-      body: Body.json(payload as any),
-    });
-    if (!response.ok) {
-      throw new Error(`Failed to get ${path}: ${response.status}`);
-    }
-    return response.data as ResponsePayload;
+    return makeTauriRequest<ResponsePayload, RequestPayload>(path, method, payload, responseParser);
+  }
+}
+
+if (import.meta.env.DEV) {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (window as any).universalRequest = universalRequest;
+}
+
+async function makeWebRequest<ResponsePayload = unknown, RequestPayload = unknown>(
+  method: HttpMethod,
+  path: string,
+  payload: RequestPayload | undefined,
+  responseParser: ResponseParser,
+) {
+  console.log(`WEB FETCH REQUEST: ${method} ${path}`);
+  const headers: HeadersInit =
+    payload instanceof FormData
+      ? {}
+      : {
+          Accept: 'application/json',
+          'Content-Type': 'application/json',
+        };
+  const response = await fetch(path, {
+    method,
+    headers,
+    body: payload instanceof FormData ? payload : payload ? JSON.stringify(payload) : undefined,
+  });
+  console.log('WEB FETCH RESPONSE STATUS:', response.status);
+  if (!response.ok) {
+    throw new Error(`Failed to ${method} ${path}: ${response.statusText}`);
+  }
+  switch (responseParser) {
+    case 'JSON':
+      return (await response.json()) as ResponsePayload;
+    case 'ArrayBuffer':
+      return (await response.arrayBuffer()) as ResponsePayload;
   }
 }
+
+async function makeTauriRequest<ResponsePayload = unknown, RequestPayload = unknown>(
+  path: string,
+  method: HttpMethod,
+  payload: RequestPayload | undefined,
+  responseParser: ResponseParser,
+) {
+  const url = REFSTUDIO_HOST + path;
+  console.log(`TAURI FETCH REQUEST: ${method} ${url}`);
+  const responseType = responseParser === 'JSON' ? ResponseType.JSON : ResponseType.Binary;
+  const response = await tauriFetch(url, {
+    method,
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-explicit-any
+    body: payload ? Body.json(payload as any) : undefined,
+    responseType,
+  });
+  console.log('TAURI FETCH RESPONSE STATUS', response.status);
+  if (!response.ok) {
+    throw new Error(`Failed to ${method} ${path}: ${response.status} - ${JSON.stringify(response.data)}`);
+  }
+  switch (responseParser) {
+    case 'JSON':
+      console.log('TAURI FETCH RESPONSE DATA', response.data);
+      return response.data as ResponsePayload;
+    case 'ArrayBuffer':
+      console.log('TAURI FETCH RESPONSE DATA', response.data);
+      return response.data as ResponsePayload;
```

Comment:
So `response.data` is already an `ArrayBuffer` or a parsed JSON object depending on the response type? Interesting API!

Reply:
I know!!